packages/playground/CoreApp/App.cpp

```diff
@@ -0,0 +1,75 @@
+#include "pch.h"
+#include "CoreApp.h"
+
+// This app demonstrates the usage of the Microsoft.ReactNative CoreApp APIs
+// use the preprocessor defines WITH_MODULES (true/false), FROM_JSON (true/false),
+// or neither to exercise the RNStartCoreAppWithModules, RNStartCoreApp, and RNStartCoreAppFromConfigJson APIs.
+#define DEFINE_MODULES
+
+// #define WITH_MODULES
+#define FROM_JSON
+
+#ifdef DEFINE_MODULES
+#include <ModuleRegistration.h>
+#include <NativeModules.h>
+#include <winrt/Microsoft.ReactNative.h>
+
+REACT_MODULE(LogBox)
+struct LogBox {
+  REACT_INIT(Initialize)
+  void Initialize(const winrt::Microsoft::ReactNative::ReactContext &) noexcept {
+    MessageBox(nullptr, L"Native Modules work!", L"Success", MB_OK);
+  }
+};
+
+struct ThisAppPackageProvider
+    : winrt::implements<ThisAppPackageProvider, winrt::Microsoft::ReactNative::IReactPackageProvider> {
+  void CreatePackage(winrt::Microsoft::ReactNative::IReactPackageBuilder const &packageBuilder) noexcept {
+    for (auto const *reg = winrt::Microsoft::ReactNative::ModuleRegistration::Head(); reg != nullptr;
+         reg = reg->Next()) {
+      packageBuilder.AddModule(reg->ModuleName(), reg->MakeModuleProvider());
+    }
+  }
+};
+
+extern "C" __declspec(dllexport) void *__cdecl MySpecialPackageProvider() {
+  auto provider = winrt::make<ThisAppPackageProvider>();
+  void *abi{nullptr};
+  winrt::copy_to_abi(provider, abi);
+  return abi;
+}
+
+#endif
+
+#ifdef FROM_JSON
+int __stdcall wWinMain(_In_ HINSTANCE, _In_opt_ HINSTANCE, _In_ PWSTR /*server*/, _In_ int /*showCommand*/) {
+  RNStartCoreAppFromConfigJson(L"app.config.json", nullptr);
+}
+#else
+int __stdcall wWinMain(_In_ HINSTANCE, _In_opt_ HINSTANCE, _In_ PWSTR /*server*/, _In_ int /*showCommand*/) {
+
+#ifdef WITH_MODULES
+  RNStartCoreAppWithModules(
+      [](RNCoreApp *app) {
+        app->componentName = L"RNTesterApp";
+        app->jsBundleFile = LR"(Samples\rntester)";
+        app->useDeveloperSupport = true;
+        app->useWebDebugger = false;
+      },
+      [](RNCoreApp *app) {
+        app->packageProvidersAbi = new void *[1];
```

Comment:
> new void *[1]; [](http://example.com/codeflow?start=34&length=15)

It is not ABI-safe. We cannot allocate memory in one DLL and deallocate it in another. Each DLL may use its onw rules how the memory is allocated and deallocated.

Reply:
where are you seeing this be deallocated?

Reply:
changed this to be COM pool, but it wasn't being deallocated AFAICT so it was fine :)

Reply:
My point was that we do not need to do any allocation at all. Having an array in the stack must work just fine in that case.
Otherwise, this sample shows how to leak memory if we never de-allocate it.

Reply:
Since it is only one package provider in this sample, we can just store a pointer to a local variable with the package provider.